File: server/module/module-system/system-server/src/service/system_post.rs
use std::str::FromStr;

use common::constants::enum_constants::{STATUS_DISABLE, STATUS_ENABLE};
use common::interceptor::orm::support_order::SupportOrder;
use sea_orm::{ActiveModelTrait, ColumnTrait, Condition, DatabaseConnection, EntityTrait, Order, PaginatorTrait, QueryFilter, QueryOrder};
use crate::model::system_post::{Model as SystemPostModel, ActiveModel as SystemPostActiveModel, Entity as SystemPostEntity, Column};
use system_model::request::system_post::{CreateSystemPostRequest, UpdateSystemPostRequest, PaginatedKeywordRequest};
use system_model::response::system_post::SystemPostResponse;
use crate::convert::system_post::{create_request_to_model, update_request_to_model, model_to_response};
use anyhow::{Result, anyhow};
use sea_orm::ActiveValue::Set;
use common::base::page::PaginatedResponse;
use common::context::context::LoginUserContext;
use common::interceptor::orm::active_filter::ActiveFilterEntityTrait;

pub async fn create(db: &DatabaseConnection, login_user: LoginUserContext, request: CreateSystemPostRequest) -> Result<i64> {
    let mut system_post = create_request_to_model(&request);
    system_post.creator = Set(Some(login_user.id));
    system_post.updater = Set(Some(login_user.id));
    system_post.tenant_id = Set(login_user.tenant_id);
    let system_post = system_post.insert(db).await?;
    Ok(system_post.id)
}

pub async fn update(db: &DatabaseConnection, login_user: LoginUserContext, request: UpdateSystemPostRequest) -> Result<()> {
    let system_post = SystemPostEntity::find_by_id(request.id)
        .one(db)
        .await?
        .ok_or_else(|| anyhow!("记录未找到"))?;

    let mut system_post = update_request_to_model(&request, system_post);
    system_post.updater = Set(Some(login_user.id));
    system_post.update(db).await?;
    Ok(())
}

pub async fn delete(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let system_post = SystemPostActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        deleted: Set(true),
        ..Default::default()
    };
    system_post.update(db).await?;
    Ok(())
}

pub async fn get_by_id(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<Option<SystemPostResponse>> {
    let condition = Condition::all()
            .add(Column::Id.eq(id))
            .add(Column::TenantId.eq(login_user.tenant_id));
            
    let system_post = SystemPostEntity::find_active_with_condition(condition)
        .one(db).await?;
    Ok(system_post.map(model_to_response))
}

pub async fn get_paginated(db: &DatabaseConnection, login_user: LoginUserContext, params: PaginatedKeywordRequest) -> Result<PaginatedResponse<SystemPostResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));
    let mut query = SystemPostEntity::find_active_with_condition(condition);

    let paginator = query
        .support_order(params.base.sort_field, params.base.sort, Some(vec![(Column::Id, Order::Asc)]))
        .paginate(db, params.base.size);

    let total = paginator.num_items().await?;
    let total_pages = (total + params.base.size - 1) / params.base.size; // 向上取整
    let list = paginator
        .fetch_page(params.base.page - 1) // SeaORM 页码从 0 开始，所以减 1
        .await?
        .into_iter()
        .map(model_to_response)
        .collect();

    Ok(PaginatedResponse {
        list,
        total_pages,
        page: params.base.page,
        size: params.base.size,
        total,
    })
}

pub async fn list(db: &DatabaseConnection, login_user: LoginUserContext) -> Result<Vec<SystemPostResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let list = SystemPostEntity::find_active_with_condition(condition)
        .all(db).await?;
    Ok(list.into_iter().map(model_to_response).collect())
}

pub async fn enable(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let system_post = SystemPostActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        status: Set(STATUS_ENABLE),
        ..Default::default()
    };
    system_post.update(db).await?;
    Ok(())
}

pub async fn disable(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let system_post = SystemPostActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        status: Set(STATUS_DISABLE),
        ..Default::default()
    };
    system_post.update(db).await?;
    Ok(())
}

File: admin/src/api/system_post.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { api } from '@/utils/request';

const apis = {
  create: '/system/system_post/create', // 新增
  update: '/system/system_post/update', // 修改
  delete: '/system/system_post/delete', // 删除
  get: '/system/system_post/get', // 单条查询
  list: '/system/system_post/list', // 列表查询
  page: '/system/system_post/page', // 分页查询
  enable: '/system/system_post/enable', // 启用
  disable: '/system/system_post/disable', // 禁用
}

export interface SystemPostRequest {
  id: number; // id
  code: string; // 职位编码
  name: string; // 职位名称
  sort: number; // 显示顺序
  status: number; // 状态（0正常 1停用）
  remark: string; // 备注
}

export interface SystemPostResponse {
  id: number; // id
  code: string; // 职位编码
  name: string; // 职位名称
  sort: number; // 显示顺序
  status: number; // 状态（0正常 1停用）
  remark: string; // 备注
  creator: number; // 创建者id
  create_time: string; // 创建时间
  updater: number; // 更新者id
  update_time: string; // 更新时间
}

export interface SystemPostQueryCondition extends PaginatedRequest {
  
}

export const createSystemPost = (system_post: SystemPostRequest): Promise<number> => {
  return api.post<number>(apis.create, system_post);
}

export const updateSystemPost = (system_post: SystemPostRequest): Promise<void> => {
  return api.post<void>(apis.update, system_post);
}

export const deleteSystemPost = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getSystemPost = (id: number): Promise<SystemPostResponse> => {
  return api.get<SystemPostResponse>(`${apis.get}/${id}`);
}

export const listSystemPost = (): Promise<Array<SystemPostResponse>> => {
  return api.get<Array<SystemPostResponse>>(apis.list);
}

export const pageSystemPost = (condition: SystemPostQueryCondition): Promise<PaginatedResponse<SystemPostResponse>> => {
  return api.get<PaginatedResponse<SystemPostResponse>>(apis.page, condition);
}

export const enableSystemPost = (id: number): Promise<void> => {
  return api.post<void>(`${apis.enable}/${id}`);
}

export const disableSystemPost = (id: number): Promise<void> => {
  return api.post<void>(`${apis.disable}/${id}`);
}